Introduce a SegmentList.to_table method

alurban opened the pull request:
This PR introduces a new method, `SegmentList.to_table`, to cast `SegmentList` objects as type `~gwpy.table.Table`. This is provided mainly to have a way to write `SegmentList` objects to standard CSV tables, which has proven useful in delivering data products to LIGO management; see [**this tool**](https://git.ligo.org/detchar/ligo-summary-pages/blob/master/scripts/gw_epoch_time_accounting) for the LIGO summary pages (requires `LIGO.ORG` credentials).

cc @duncanmmacleod

duncanmmacleod commented on gwpy/segments/tests/test_segments.py:
```diff
@@ -82,6 +84,15 @@ def test_coalesce(self):
+        assert_array_equal(segtable['end'], [seg[1] for seg in segmentlist])
+        assert_array_equal(
+            segtable['duration'], [abs(seg) for seg in segmentlist])
+
```
Can you just instead use `assert_table_equal` and build the expected table on the fly? Something like:

```python
    def test_to_table(self, segmentlist):
        segtable = segmentlist.to_table()
        assert_table_equal(
            segtable,
            Table(
                rows=[
                    (0, 1, 2, 1),
                    (1, 3, 4, 1),
                    (2, 4, 6, 2),
                    (3, 8, 10, 2),
                ],
                names=('index', 'start', 'end', 'duration'),
            ),
        )
```
(untested)

alurban replied:
@duncanmmacleod, your proposed unit test passes (and runs about 6.5 times faster), so I've implemented it in the latest commit.

duncanmmacleod commented on gwpy/segments/segments.py:
```diff
@@ -144,6 +144,25 @@ def coalesce(self):
     coalesce.__doc__ = segmentlist.coalesce.__doc__
 
+    def to_table(self):
+        """Convert this `SegmentList` to a `~gwpy.table.Table`
```
this class is really just `astropy.table.Table`
```suggestion
        """Convert this `SegmentList` to a `~astropy.table.Table`
```

duncanmmacleod commented on gwpy/segments/segments.py:
```diff
@@ -144,6 +144,25 @@ def coalesce(self):
+        objects in comma-separated value (CSV) format, via the
+        :meth:`~gwpy.table.Table.write` method.
+        """
+        from ..table import Table
```
```suggestion
        from astropy.table import Table
```

duncanmmacleod commented on gwpy/segments/tests/test_segments.py:
```diff
@@ -25,7 +25,9 @@
 import h5py
 
-from ...testing.utils import (assert_segmentlist_equal, TemporaryFilename)
+from ...testing.utils import (assert_array_equal, assert_segmentlist_equal,
```
```suggestion
from ...testing.utils import (assert_segmentlist_equal, assert_table_equal,
```

duncanmmacleod commented on gwpy/segments/tests/test_segments.py:
```diff
@@ -25,7 +25,9 @@
-from ...testing.utils import (assert_segmentlist_equal, TemporaryFilename)
+from ...testing.utils import (assert_array_equal, assert_segmentlist_equal,
+                              TemporaryFilename)
+from ...table import Table
```
```suggestion
from astropy.table import Table
```
(with appropriate repositioning)

alurban commented:
Thanks for your review @duncanmmacleod, I've addressed your comments and moved the method-level import of `gwpy.table.Table` out into a _module_-level import of `astropy.table.Table`, since it no longer causes a circular import.

duncanmmacleod commented on gwpy/segments/segments.py:
```diff
@@ -144,6 +145,24 @@ def coalesce(self):
+            [seg[0] for seg in self],
+            [seg[1] for seg in self],
+            [abs(seg) for seg in self],
+        ], names=('index', 'start', 'end', 'duration'))
```
Looking at this again, does it make more sense to only unpack the segmentlist once:

```python
        return Table(
            list(zip(*((i, s[0], s[1], abs(s)) for i, s in enumerate(a)))),
            names=('index', 'start', 'end', 'duration'),
        )

alurban replied:
@duncanmmacleod, this speeds up the unit test by another factor of 2 (and takes up fewer lines), so I went ahead and pushed it with the last commit.

duncanmmacleod commented:
@alurban, the circleci-based conda builds are failing, the only thing I can thing of to try is to refresh the circleci cache:

```patch
diff --git a/.circleci/config.yml b/.circleci/config.yml
index 2a9973e0..84765bbf 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -82,7 +82,7 @@ aliases:
         - checkout
         - *attach_workspace
         - restore_cache:
-            key: v4-gwpy-{{ .Environment.CIRCLE_JOB }}
+            key: v5-gwpy-{{ .Environment.CIRCLE_JOB }}
         - run: *set_python_environment
         - run:
             name: Install
@@ -90,7 +90,7 @@ aliases:
         - run: *run-tests
         - run: *codecov
         - save_cache:
-            key: v4-gwpy-{{ .Environment.CIRCLE_JOB }}
+            key: v5-gwpy-{{ .Environment.CIRCLE_JOB }}
             paths:
               - "/opt/conda/envs/gwpyci"
         - *store_test_results
```

alurban commented:
@duncanmmacleod, I'm giving it a shot, here goes nothing...

alurban commented:
@duncanmmacleod, this gambit was successful, thanks for the suggestion
